tests: Add a new GUI testing framework

Elv13 opened the pull request:
This framework allow widgets and user interface related code to be tested.

This should close the gap in our testing pipeline. It sits between unit testing and the full integration testing.

#### It has the following features:

 * Generate documentation examples
 * Generate SVG screenshots of widgets
 * Generate dynamic ldoc content
 * Emulate just enough of the C API for the tests to run
 * Tests corner cases like weird multi screen that are harder to test with Xephyr
 * provide a template mechanism to generate different kind of images. Each module can have their own template or inherit from the upper module.

#### Drawbacks:

 * With a thousand tests, the build get very slow, so should only be enabled on the CI. I still have to test parallel test generation. Given the bottleneck is mostly cmake and fork/pipe, I could technically generate a test list, then run 16 in parallel. With the 60 tests I already wrote, it add ~7 seconds to the build.
 * It is currently 1 test per file. This cause tons of files to be created. The reason behind this is to clean the state at every run. For example, tests can set random virtual screen setup, but I don't want to have to write code to rollback everything.

#### WIP:

 * Eventually run some regression testing on the SVG files. A simple vector path comparison should detect regressions/changes. A github bot could report them in the pull request with before and after images in the comment.

@blueyed Do you have any feature you want in this before I get too far?

Elv13 commented:
Here are pdf (print) output of the `widget` template  and the `screen+client+mouse` template.

[wibox_widget_background.pdf](https://github.com/awesomeWM/awesome/files/185708/wibox_widget_background.pdf)
[awful_placement.pdf](https://github.com/awesomeWM/awesome/files/185709/awful_placement.pdf)

Please ignore all the new code in these files, it is work in progress, full of typo and there is good reasons why the placement rewrite design is a little weird.

The screen template is not done yet, it is missing screen, mouse and client text labels and visual indicator of the initial states versus the result.

blueyed commented on awesomeConfig.cmake:
```diff
@@ -299,6 +299,11 @@ set(AWESOME_THEMES_PATH      ${AWESOME_DATA_PATH}/themes)
     # Generate some images and examples
     include(docs/generate_examples.cmake)
+
+    # Use `include` rather than `add_subdirectory` to keep the variables
```
Punctuation!
Don't need to change it now, but please consider it in the future!

blueyed commented on awesomeConfig.cmake:
```diff
@@ -299,6 +299,11 @@ set(AWESOME_THEMES_PATH      ${AWESOME_DATA_PATH}/themes)
+
+    # Use `include` rather than `add_subdirectory` to keep the variables
+    # The file is a valid CMakeLists.txt and can be executed directly if only
+    # the images artefacts are needed.
```
s/images/image/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# Module is designed to allow the Awesome documentation examples to be tested.
```
s/Module/This module/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# Module is designed to allow the Awesome documentation examples to be tested.
+#
+# It shim enough of Awesome C API to allow code to be executed without an
```
s/shim/shims/

blueyed replied:
s/of Awesome/of the Awesome/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# Module is designed to allow the Awesome documentation examples to be tested.
+#
+# It shim enough of Awesome C API to allow code to be executed without an
+# actual X server or running Awesome process. These test are not genuine
```
s/test/tests/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# actual X server or running Awesome process. These test are not genuine
+# integration tests, but they are the next best thing.
+#
+# As secondary goals, this module also generate images of the test result where
```
s/generate/generates/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# integration tests, but they are the next best thing.
+#
+# As secondary goals, this module also generate images of the test result where
+# relevant. Those images are used by the documentation and help developpers
```
s/developpers/developers/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+#
+# As secondary goals, this module also generate images of the test result where
+# relevant. Those images are used by the documentation and help developpers
+# track user interface regressions and glitches. Finally, it also help to find
```
s/help/helps/

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# broken code.
+cmake_minimum_required(VERSION 3.0.0)
+
+# Get and update the LUA_PATH so the scripts can be executed without Awesome
```
dot-at-the-end.

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+# Get and update the LUA_PATH so the scripts can be executed without Awesome
+execute_process(COMMAND lua -e print\(package.path\) OUTPUT_VARIABLE "LUA_PATH_")
+
+# Add the main awesome lau libraries
```
Another missing dot.

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+execute_process(COMMAND lua -e print\(package.path\) OUTPUT_VARIABLE "LUA_PATH_")
+
+# Add the main awesome lau libraries
+# set(${LUA_PATH_} "${SOURCE_DIR}/lib/?;${SOURCE_DIR}/lib/?.lua;${LUA_PATH_}")
```
Remove the commented code?

blueyed replied:
Or is it useful for un-commenting locally?

blueyed commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,243 @@
+
+# Add the C API shims
+# set(${LUA_PATH_} "${SOURCE_DIR}/lib/?;${SOURCE_DIR}/lib/?.lua;${LUA_PATH_}")
+set(LUA_PATH3_ "\
```
Same as above.

Elv13 commented:
Updated, I also added some of the other WIP commits. The shims are currently terrible and full of copy pasted hacks. It's good enough to develop with, but it will have to be cleaned up before merging.

The shims can not also emulate tags enough to play with them, as long as you only have 1 per virtual screen (enough for testing most of awful code paths). Other than that, it is still an early WIP.

Elv13 commented:
[tile.pdf](https://github.com/awesomeWM/awesome/files/189197/tile.pdf)

The new tile layout documentation (WIP)

Elv13 commented:
I removed the `[WIP]` tag. So here it is. The shims layer is a little messy, but good enough to do its job. It doesn't have to be perfect now as long as it allow the tests to run. This PR include tests for the background widget as an example. Coverall seem to work. The CMake part is good enough, even if a little slow. About that, it is still faster than the Xephyr tests, so I guess it is fine.

@psychon I know this is a big, scary and far reaching. The logic behind investing time into this to make it happen is for what it will eventually bring. The documentation improvement are nice, but being able to cover parts of that code that was previously very hard to cover is what will eventually bring Awesome forward. Yes, all those tests represent *a lot* of code. It is not fun to write, I will tell you that, but overall I am convinced it is a step in the right direction.

The new `awful.placement` and my new client layout system are tested using this, so it has to be merged so I can advance on the other fronts.

psychon commented on lib/gears/surface.lua:
```diff
@@ -205,6 +206,52 @@ function surface.apply_shape_bounding(draw, shape, ...)
+    local h = hierarchy.new({dpi=96}, self, width, height, redraw, main_layout, {})
+
+    main_layout(h)
+    h:_redraw()
```
I'd replace this with `redraw(h)` instead. However, I don't think this method is necessary (see below).

psychon commented on lib/gears/surface.lua:
```diff
@@ -205,6 +206,52 @@ function surface.apply_shape_bounding(draw, shape, ...)
+    local img = cairo.SvgSurface.create(path, width, height)
+    local cr = cairo.Context(img)
+
+    return run_in_hierarchy(widget, cr, width, height)
```
Replace this with `hierarchy.new({dpi=96}, widget, width, height, function() end, function() end, {}):draw({dpi=96}, cr)` (for extra niceness, you can make and use a `local function no_op() end` and move the context into a variable, too).

There will be no relayouts necessary here and no redraws either (in fact, redraws would be bad, because the widget would be drawn to the SVG twice).

(Same thing below)

psychon replied:
Also, add `img:finish()` at the end (only then is the SVG file actually written, right now it depends on the garbage collector to finish the SVG Surface)

Elv13 replied:
Finish was already in one of the template, now, both have it. As for removing the local function, I prefer not, I don't think it cover all corner cases such as creating new images for each "frame" of the main loop. It is currently not implemented, but as it would add some (~5 lines) of new logic, the code duplication would start to be more problematic. Also, there is a PNG version too. I just didn't add it in this commit as it was broken and I have no tests that require bitmap output yet.

psychon commented on lib/gears/surface.lua:
```diff
@@ -205,6 +206,52 @@ function surface.apply_shape_bounding(draw, shape, ...)
+function surface.widget_to_surface(widget, width, height, format)
+    local img = cairo.ImageSurface(format or cairo.Format.ARGB32, width, height)
+    local cr = cairo.Context(img)
+    local h = run_in_hierarchy(widget, cr, width, height)
```
Ok, perhaps the code suggested above should be moved into a helper function. Something like `layout_and_draw_widget(widget, cr, width, height)`?

psychon commented on lib/gears/surface.lua:
```diff
@@ -205,6 +206,52 @@ function surface.apply_shape_bounding(draw, shape, ...)
+-- @tparam widget widget A widget
+-- @tparam number width The surface width
+-- @tparam number height The surface height
+-- @param[opt=cairo.Format.ARGB32] format The surface format
```
Return values are undocumented.

Elv13 replied:
done

psychon commented on tests/examples/shims/awesome.lua:
```diff
@@ -0,0 +1,27 @@
+awesome:add_signal("spawn::canceled")
+awesome:add_signal("spawn::timeout")
+
+return awesome
```
Modeline is missing

psychon commented on tests/examples/shims/client.lua:
```diff
@@ -0,0 +1,202 @@
+
+local clients = {}
+
+--TODO merge all other implementations
```
How about moving this to `awesome._shim_convert_singleton`?

psychon commented on tests/examples/shims/client.lua:
```diff
@@ -0,0 +1,202 @@
+    function obj.connect_signal(name, func)
+        return obj._connect_signal(obj, name, func)
+    end
+    obj.weak_connect_signal = obj.connect_signal
```
The C-API doesn't have weak signal connections, so I think this part can be dropped.

psychon commented on tests/examples/shims/client.lua:
```diff
@@ -0,0 +1,202 @@
+        if new then
+            for k,v in pairs(new) do
+                ret:emit_signal("property::"..k, v)
+                ret[k] = v
```
I think the order is wrong, you should first change the value and then emit the signal (and only emit if the value actually changed)

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -0,0 +1,251 @@
+    # Execute the script, leave the image extension decision to the test
+    # SVG is preferred, but PNG is better suited for some tests, like bitmap
+    # patterns.
+    execute_process(
```
Should there be a `message(STATUS "Doing foo")` here, so that we know that it is executing something like this? If not, should there be such a message at the beginning and at the end of the file so that one can at least see how much time is spent here in the build?

Elv13 replied:
done

psychon commented on tests/examples/awful/template.lua:
```diff
@@ -0,0 +1,159 @@
+local pangocairo = require("lgi").PangoCairo
+
+-- Set the global shims
+-- luacheck: globals awesome
```
I think it is enough to do `luacheck: globals awesome root tag screen client mouse`, it applies from "here" to the end of the file (unless the comment is on a not-otherwise-empty line)

psychon commented on tests/examples/wibox/template.lua:
```diff
@@ -0,0 +1,63 @@
+        surface["widget_to_"..image_type](container, image_path.."."..image_type, w or f_w, h or f_h)
+
+        -- Quit normally, else the .svg are going to be empty
+        main_loop:quit()
```
This will be fixed by my `:finish()` suggestion on the commit that adds these methods

psychon commented on tests/examples/wibox/template.lua:
```diff
@@ -0,0 +1,63 @@
+local counter = 1
+
+-- Loop as fast as possible
+GLib.timeout_add_seconds(0, 0,function()
```
Can you try using `add_idle` (I'm not totally sure about the name). This function will be called whenever there is nothing else to do, which seems better than the 0 second timeout.

Actually... does anything in awesome-lua-land use the GLib main loop? I think you can replace all of this with
```
for counter = 1, 10 do
  awesome:emit_signal("refresh")
end
```
here comes the code you do on the 10th iteration

psychon commented:
Thanks for working on this. You are right that this looks a bit scary, but I'll live with it. My comments above are mostly minor things.

psychon commented on tests/examples/wibox/template.lua:
```diff
@@ -0,0 +1,63 @@
+        f_w, f_h = f_w+10, f_h+10
+
+        -- Save to the output file
+        surface["widget_to_"..image_type](container, image_path.."."..image_type, w or f_w, h or f_h)
```
There is only `widget_to_svg`, or do I miss something?

Elv13 replied:
There is the png version too, it's not in this PR, but it will come when I will need it. This PR is already 1k line with only the code I need...

(there is no huge second part, don't worry ;) ).

Elv13 commented:
Updated. I only partially ignored the `gears.surface` comments. By the way, the SVG is not written twice, it is written on `:finish()` and the "CLEAR" operator do it's job, so it is not written on top of itself.

psychon commented:
Thanks. And yeah, you are right, drawing the background should handle multiple frames. It's still unneccessary here to handle re-layouts and re-draws, but if you insist...

psychon commented on lib/gears/surface.lua:
```diff
@@ -205,6 +206,55 @@ function surface.apply_shape_bounding(draw, shape, ...)
 
+local function run_in_hierarchy(self, cr, width, height)
+    local function main_layout(h)
+        h:update({dpi=96}, self, width, height)
```
Hmm. This doesn't handle redraws due to relayouts, does it? That part of the code works with a `cairo.Region` that is returned from `:update()`.

Elv13 replied:
Your better placed than me to know your code, but so far it worked fine

psychon replied:
Because no widget can emit any signals that this hierarchy should react to, so this code is never actually executed. ;-)
As I said, for the code you have right now, you could just use no-op-functions for these callbacks.

Elv13 replied:
Are you sure? I am just asking, but isn't `widget::layout_changed` supposed to end up calling this? If I create an `.svg` at main loop cycle `#1`, then do `:finish()` at cycle `#10` after chaning the widget 9 time, isn't it better to have this called? If not, then I will remove it, test and merge this PR as it is the only remaining issue.

psychon replied:
Hm... apparently I misunderstood your intention with this function.
In the following code, I'd espect a SVG saying "foo" and not "bar":
```
local t = wibox.widget.textbox("foo")
gears.surface.widget_to_svg(t, "/tmp/t.svg", 100, 100)
t:set_text("bar")
```
Apparently, you want it to say `bar`. And mentioned something like "multiple frames" being produced. I don't really get this.

Elv13 replied:
nevermind, one way or another, it is unused. I will remove it for now. Cairo doesn't yet support multiframe SVG anyway.

Elv13 commented:
Umm, the examples seem to be processed, but not kept by the awesome-robot. I probably missed something, I will check that tomorrow. It's probably minor as it work locally with the same script.
